src/test/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecutionTest.java

```diff
@@ -250,4 +254,78 @@ public boolean start() throws Exception {
         });
     }
 
-}
\ No newline at end of file
+    /**
+     * When an exception is thrown in {@link BodyExecutionCallback#onStart(StepContext)}, we must ensure that we clean
+     * up after ourselves and not leave a zombie work unit runing on an executor.
+     * @throws Exception
+     */
+    @Test
+    public void unhandledAssertionsShouldNotCreateZombieExecutions() throws Exception {
+        rr.then(r -> {
+            r.jenkins.setNumExecutors(1);
+            WorkflowJob job = r.createProject(WorkflowJob.class);
+            job.setDefinition(new CpsFlowDefinition("node('master') { withStartFailure { echo 'oh dear' } }"));
+            r.buildAndAssertStatus(Result.FAILURE, job);
+            assertThat(r.jenkins.getComputers()[0].getExecutors().get(0).getCurrentWorkUnit(), nullValue());
+        });
+    }
+
+    public static class WithStartFailureStep extends Step {
+
+        @DataBoundConstructor
+        public WithStartFailureStep() {}
+
+        @TestExtension("unhandledAssertionsShouldNotCreateZombieExecutions")
+        public static class DescriptorImpl extends StepDescriptor {
+            @Override
+            public String getFunctionName() {
+                return "withStartFailure";
+            }
+
+            @Override
+            public Set<? extends Class<?>> getRequiredContext() {
+                return Collections.emptySet();
+            }
+
+            @Override
+            public boolean takesImplicitBlockArgument() {
+                return true;
+            }
+        }
+
+        @Override
+        public StepExecution start(StepContext context) throws Exception {
+            return new WithStartFailureStepExecution(context);
+        }
+
+        static class WithStartFailureStepExecution extends AbstractStepExecutionImpl {
+
+            WithStartFailureStepExecution(final StepContext context) {
+                super(context);
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

Reply:
This is a constructor, not a regular method.

Reply:
noticed.